python/ray/train/base_trainer.py

```diff
@@ -164,8 +176,169 @@ def __init__(
+        )
+        trainer_state_path = cls._maybe_sync_down_trainer_state(path)
+        assert (
+            trainer_state_path.exists()
+        ), f"Did not find trainer state at {str(trainer_state_path)}."
+
+        with open(trainer_state_path, "rb") as fp:
+            original_trainer = pickle.load(fp)
+        assert type(original_trainer) == cls, (
+            f"Invalid trainer type. Cannot restore a trainer of type "
+            f"{type(original_trainer)} with `{cls.__name__}.restore`. "
+            f"Use `{type(original_trainer).__name__}.restore` instead."
+        )
+
+        # Get the param dict used to initialize the original trainer
+        param_dict = original_trainer._param_dict
+
+        original_datasets = original_trainer.datasets or {}
+        if original_datasets and not datasets:
+            raise ValueError(
+                "The following datasets need to be provided again on restore: "
+                f"{list(original_datasets.keys())}\n"
+                f"Use {cls.__name__}.restore(..., datasets=datasets) "
+                "with the datasets that were provided to the original trainer."
+            )
+        datasets = datasets or {}
+        assert set(original_datasets.keys()) == set(datasets.keys()), (
+            "The provided datasets don't match the original dataset keys.\n"
+            f"  Expected datasets for the keys: {list(original_datasets.keys())}\n"
+            f"  Actual datasets provided: {list(datasets.keys())}"
+        )
+        param_dict["datasets"] = datasets
+
+        # If no preprocessor is re-specified, then it will be set to None
+        # here and loaded from the latest checkpoint
+        param_dict["preprocessor"] = preprocessor
+
+        if scaling_config:
+            param_dict["scaling_config"] = scaling_config
+
+        for param_name, val in kwargs.items():
+            # Overwrite the old value if something is passed into restore
+            if val is not None:
+                param_dict[param_name] = val
+
+        trainer = cls(**param_dict)
+        trainer._restore_path = path
+        return trainer
+
+    @classmethod
+    def can_restore(cls: Type["BaseTrainer"], path: Union[str, Path]) -> bool:
+        """Checks whether a given directory contains a restorable Train experiment.
+
+        Args:
+            path: The path to the experiment directory of the Train experiment.
+                This can be either a local directory (e.g. ~/ray_results/exp_name)
+                or a remote URI (e.g. s3://bucket/exp_name).
+
+        Returns:
+            bool: Whether or not this path exists and contains the pickled Trainer
+        """
+        path = str(path)
+        if is_non_local_path_uri(path):
+            dir_contents = list_at_uri(path)
+        else:
+            dir_contents = [] if not os.path.exists(path) else os.listdir(path)
+        return bool(dir_contents) and _TRAINER_PKL in dir_contents
```

Comment:
```suggestion
        return dir_contents and _TRAINER_PKL in dir_contents
```

BTW. this is the same as in the other PR, right? Should we make it a separate AIR-level function that is simply wrapped around by Trainer & Tuner classmethods?

Reply:
Yes, can do.